scripts/core/editor3/reducers/suggestions.jsx

```diff
@@ -0,0 +1,360 @@
+import {EditorState, Modifier, RichUtils} from 'draft-js';
+import {onChange} from './editor3';
+import {getEntityKeyByOffset, getCharByOffset,
+    getEntityKey, getEntity, setEntity, deleteEntity} from './utils';
+
+const suggestions = (state = {}, action) => {
+    switch (action.type) {
+    case 'TOGGLE_SUGGESTIONS_MODE':
+        return toggleSuggestionsMode(state);
+    case 'CREATE_ADD_SUGGESTION':
+        return createAddSuggestion(state, action.payload);
+    case 'CREATE_DELETE_SUGGESTION':
+        return createDeleteSuggestion(state, action.payload);
+    default:
+        return state;
+    }
+};
+
+/**
+ * @ngdoc method
+ * @name toggleSuggestionsMode
+ * @param {Object} state
+ * @return {Object} returns new state
+ * @description Disable/enable the suggestions mode.
+ */
+const toggleSuggestionsMode = (state) => {
+    const {suggestionsMode, editorState} = state;
+    const newState = resetDefaultSugestionStyle(editorState);
+
+    return {
+        ...onChange(state, newState),
+        suggestionsMode: !suggestionsMode
+    };
+};
+
+/**
+ * @ngdoc method
+ * @name createAddSuggestion
+ * @param {Object} state
+ * @param {String} text - the suggestion added text
+ * @param {Object} data - info about the author of suggestion
+ * @return {Object} returns new state
+ * @description Add a new suggestion of type ADD.
+ */
+const createAddSuggestion = (state, {text, data}) => {
+    let {editorState} = state;
+    const selection = editorState.getSelection();
+    const length = selection.getEndOffset() - selection.getStartOffset();
+    let i;
+
+    if (length !== 0) {
+        // if text is selected, apply delete suggestion for every selected character
+        editorState = changeEditorSelection(editorState, length, 0, false);
+        for (i = 0; i < length; i++) {
+            editorState = setDeleteSuggestionForCharacter(editorState, data);
+        }
+    }
+
+    for (i = 0; i < text.length; i++) {
+        // for every character from inserted text apply add suggestion
+        editorState = setAddSuggestionForCharacter(editorState, text[i], data);
+    }
+
+    return onChange(state, editorState, true);
+};
+
+/**
+ * @ngdoc method
+ * @name createDeleteSuggestion
+ * @param {Object} state
+ * @param {Object} data - info about the author of suggestion
+ * @return {Object} returns new state
+ * @description Add a new suggestion of type DELETE.
+ */
+const createDeleteSuggestion = (state, {data}) => {
+    let {editorState} = state;
+    const selection = editorState.getSelection();
+    const length = selection.getEndOffset() - selection.getStartOffset();
+    let i;
+
+    if (length === 0) {
+        editorState = setDeleteSuggestionForCharacter(editorState, data);
+    } else {
+        // if text is selected, apply delete suggestion for every selected character
+        editorState = changeEditorSelection(editorState, length, 0, false);
+        for (i = 0; i < length; i++) {
```

Comment:
Declare a local iterator instead of reusing the one in the upper scope. This is not a big deal in this case, but it's a good practice not to use variables from upper scopes when you don't have to.